[Documentation] Document test process

VWoeltjen opened the pull request:
Adds/moves documentation of our existing test processes to open source (complementing the Development Cycle document, which describes _when_ but not _what_)

This provides a baseline for WTD-1573, acceptance testing in WARP. 

Additionally, provides a location for adding test procedures (to document _how_) as requested in WTD-1585

Omitting Author Checklist since changes are documentation-only.

akhenry commented:
This looks good. I have two questions
1. When are the procedures details in the procedures document going to be performed. Is it at the user testing stage? If so, is it for sprints, releases, or both?
2. Where are the test procedures going to be captured? (probably doesn't need a reference in the procedures document, I'm just curious)

VWoeltjen commented:
> 1. When are the procedures detailed in the procedures document going to be performed. Is it at the user testing stage? If so, is it for sprints, releases, or both?

Thinking we'll need to section this out as we go - primarily I'm thinking of user testing. There are cases (like testathon setup) which aren't exactly "user testing" but can be sectioned out.

As to sprints/releases, I think we should be performing a subset of the user test procedures per-sprint (cherry-picked based on work from that sprint), and the full set of user test procedures per-release. I'll add some text to the Test Plan asserting this.

> 2. Where are the test procedures going to be captured? (probably doesn't need a reference in the procedures document, I'm just curious)

Currently, we're doing that on relevant issues. I'll add a note reflecting this.

I think that's the wrong way to capture user test results (or it will be once we start formalizing procedures), but for purposes of this PR I want to get to a point where we've documented our current process so that I can send a link around and solicit change requests.

VWoeltjen commented on CONTRIBUTING.md:
```diff
@@ -292,6 +292,7 @@ checklist.)
 2. Unit tests included and/or updated with changes?
 3. Command line build passes?
 4. Expect to pass code review?
+5. Changes have been smoke-tested?
```
In the interest of keeping this checklist short, I'm thinking of keeping the new smoke-testing check and removing "Expect to pass code review?" The latter question just serves as a reminder to think about code style etc, but I don't think it really adds anything to the process to have this (the reviewer will ultimately be responsible for determining if those standards are appropriately adhered-to)

akhenry replied:
That sounds reasonable to me.

larkin replied:
I also believe that code style should be enforced by an automated tool so we don't have to apply it ourselves. Wastes time in reviews.

Jslint currently fills the gap but I think we should eventually transition to jscs and eslint.  More configurable.

Pete

On Dec 2, 2015, at 11:14 AM, akhenry <notifications@github.com<mailto:notifications@github.com>> wrote:


In CONTRIBUTING.md<https://github.com/nasa/openmctweb/pull/349#discussion_r46460030>:

> @@ -292,6 +292,7 @@ checklist.)
>  2. Unit tests included and/or updated with changes?
>  3. Command line build passes?
>  4. Expect to pass code review?
> +5. Changes have been smoke-tested?


That sounds reasonable to me.

—
Reply to this email directly or view it on GitHub<https://github.com/nasa/openmctweb/pull/349/files#r46460030>.

VWoeltjen commented:
Followed up on notes/discussion from review:

* Added clarification on difference between per-sprint and per-release user testing (although this clarification is not particularly useful without specific user test cases.)
* Added some detail to test procedures (to reflect current testathon process, with deployment-specific details factored out)
* Reduced number of Author Checklist questions down to four

akhenry commented:
Looks good!

Omitting reviewer checklist because documentation only.
